src/components/VaultRow/lib/VaultDepositForm.tsx

```diff
@@ -141,17 +185,21 @@ export const VaultDepositForm: FC<VaultProps> = (props) => {
     overrides: { value: inputIsEth ? value : BigNumber.from(0) },
   })
   const depositUnderlying = useContractWrite(prepareDepositUnderlying.config)
+  const depositUnderlyingTxHash = depositUnderlying.data?.hash
   const waitDepositUnderlying = useWaitForTransaction({
-    hash: depositUnderlying.data?.hash,
-    onSuccess: onDepositSuccess,
+    hash: depositUnderlyingTxHash,
+    onSuccess: () => onDepositSuccess(depositUnderlyingTxHash),
+    onError: () => onDepositError(depositUnderlyingTxHash),
   })
 
   // Form submit handler
   const onSubmitForm: SubmitHandler<TokenFormValues> = async () => {
     if (requiresApproval) {
+      onApproveStart()
```

Comment:
Can't use onSettled method, onSuccess is fired before it. I workaround it by calling the loading here.

Reply:
There is a cleaner way to handle this without `useState` by invoking `writeAsync`:

```ts
const waitingForSigner = toastManager.loading("Waiting for signature")
await approve
  .writeAsync?.()
  .then((receipt) =>
    // this fires after the transaction has been broadcast successfully
    toastManager.loading(
      "Waiting for transaction confirmation",
      receipt.hash
    )
  )
  .catch((err) =>
    // this fires after a failure to broadcast the transaction
    toastManager.error(
      err instanceof UserRejectedRequestError
        ? "User rejected request"
        : "Error broadcasting transaction"
    )
  )
  .finally(() => toastManager.dismiss(waitingForSigner))
```

In the case of "waiting for signature" toasts, we don't have a hash so they must be manually dismissed -- but for toasts we *do* have a hash for, I really think you should try self-dismissal within the `LoadingToast` component. 

It is 100% worth implementing it this way as to remove any need to manually keep track of toast IDs and grealy streamline the entire process for us while implementing.

You'll need to accept the toast `id` and `txHash` as props on `LoadingToast` and wait for the transaction there:

```ts
// this is all you need to do; hook will not run if `hash` is undefined
useWaitForTransaction({ hash, onSettled: () => toast.dismiss(id) })
```